dask_ml/model_selection/_hyperband.py

```diff
@@ -0,0 +1,546 @@
+from copy import deepcopy
+import logging
+import math
+
+import numpy as np
+from sklearn.base import clone
+from sklearn.model_selection import ParameterSampler
+from sklearn.utils import check_random_state
+from sklearn.metrics.scorer import check_scoring
+from tornado import gen
+import toolz
+
+import dask.array as da
+from dask.distributed import as_completed, default_client, futures_of
+from distributed.metrics import time
+from dask_searchcv.model_selection import DaskBaseSearchCV
+
+from ..wrappers import Incremental
+from ._split import train_test_split
+from .._utils import copy_learned_attributes
+
+
+logger = logging.getLogger(__name__)
+
+
+def _get_hyperband_params(R, eta=3):
+    """
+    Arguments
+    ---------
+    R : int
+        The maximum number of iterations desired.
+    Returns
+    -------
+    N : list
+        The number of models for each bracket
+    R : list
+        The number of iterations for each bracket
+    brackets : list
+        The bracket identifier.
+    """
+    s_max = math.floor(math.log(R, eta))
+    B = (s_max + 1) * R
+
+    brackets = list(reversed(range(int(s_max + 1))))
+    N = [math.ceil(B / R * eta**s / (s + 1)) for s in brackets]
+    R = [int(R * eta**-s) for s in brackets]
+    return list(map(int, N)), list(map(int, R)), brackets
+
+
+def _partial_fit(model_and_meta, X, y, meta=None, fit_params={}):
+    """
+    Call partial_fit on a classifiers with X and y
+
+    Arguments
+    ---------
+    model_and_meta : tuple, (model: any, meta: dict)
+        model needs to support partial_fit. meta is assumed to have keys
+        ``mean_copy_time, iterations, partial_fit_calls, mean_fit_time``.
+        partial_fit will be called on the model until
+        ``meta['iterations'] >= meta['partial_fit_calls']``
+    X, y : np.ndarray, np.ndarray
+        Training data
+    meta : dict
+        If present, replace ``model_and_meta[1]`` with this object
+    fit_params : dict
+        Keyword args to pass to partial_fit
+
+    Returns
+    -------
+    model : any
+        The model that has been fit.
+    meta : dict
+        A new dictionary with updated information.
+
+    This function does not modify any item in place.
+
+    """
+    start = time()
+    model = deepcopy(model_and_meta[0])
+    if meta is None:
+        meta = deepcopy(model_and_meta[1])
+    else:
+        meta = deepcopy(meta)
+    meta['mean_copy_time'] += time() - start
+    while meta['iterations'] < meta['partial_fit_calls']:
+        # TODO: handle this unwrapping elsewhere
+        if isinstance(model, Incremental):
+            model.estimator.partial_fit(X, y, **fit_params)
```

Comment:
See also #236

Reply:
I'll rebase with that.

Reply:
If its easier feel free to merge from master.  We'll be squashing everything anyway.